File: python/nodes/stop_node.py
from typing import Dict, Any, Protocol
from ambler import Edges, Node, Next

class Utils(Protocol):
    def log(self, message: str) -> None:
        ...

class DefaultUtils:
    def log(self, message: str) -> None:
        print(message)

def create(
    edges: Edges[str, str],
    utils: Utils = DefaultUtils()
) -> Node[Dict[str, Any], str]:
    def node_fn(state: Dict[str, Any], key: str) -> Next[Dict[str, Any], str]:
        utils.log(f"Final count: {state['count']}")
        utils.log("Counting process terminated.")
        return edges["onDone"], state

    return node_fn

File: typescript/nodes/stopNode.ts
import { Edges, Node } from "../ambler.ts";

export interface Utils {
  log: (message: string) => void;
}

export type Edge = "onDone";

export function create<S extends { count: number }, K extends string>(
  edges: Edges<Edge, K>,
  utils: Utils = { log: console.log },
): Node<S, K> {
  return (state) => {
    utils.log(`Final count: ${state.count}`);
    utils.log("Counting process terminated.");
    return [edges.onDone, state];
  };
}
